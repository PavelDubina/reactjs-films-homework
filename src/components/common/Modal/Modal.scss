/*
 * Copyright Â© 2020 EPAM Systems, Inc. All Rights Reserved. All information contained herein is, and remains the
 * property of EPAM Systems, Inc. and/or its suppliers and is protected by international intellectual
 * property law. Dissemination of this information or reproduction of this material is strictly forbidden,
 * unless prior written permission is obtained from EPAM Systems, Inc
 */

.modal {
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: rgba(0, 0, 0, 0.8);
}

.frame {
  width: 70vw;
  height: 45vw;
}

.not_found {
  font-size: 50px;
  color: #888;
}
