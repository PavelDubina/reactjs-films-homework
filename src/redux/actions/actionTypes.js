/*
 * Copyright Â© 2020 EPAM Systems, Inc. All Rights Reserved. All information contained herein is, and remains the
 * property of EPAM Systems, Inc. and/or its suppliers and is protected by international intellectual
 * property law. Dissemination of this information or reproduction of this material is strictly forbidden,
 * unless prior written permission is obtained from EPAM Systems, Inc
 */

export const SEARCH_MOVIES_PENDING = 'SEARCH_MOVIES_PENDING'
export const SEARCH_MOVIES_SUCCESS = 'SEARCH_MOVIES_SUCCESS'
export const SEARCH_MOVIES_ERROR = 'SEARCH_MOVIES_ERROR'
export const SEARCH_GENRES_SUCCESS = 'SEARCH_GENRES_SUCCESS'
export const SEARCH_GENRES_ERROR = 'SEARCH_GENRES_ERROR'
export const CHANGE_CATEGORY = 'CHANGE_CATEGORY'
export const CHANGE_GENRE = 'CHANGE_GENRE'
export const SEARCH_DETAILS_SUCCESS = 'SEARCH_DETAILS_SUCCESS'
export const SEARCH_DETAILS_ERROR = 'SEARCH_DETAILS_ERROR'
export const CLOSE_MOVIE = 'CLOSE_MOVIE'
export const SEARCH_TRAILER_SUCCESS = 'SEARCH_TRAILER_SUCCESS'
export const SEARCH_TRAILER_ERROR = 'SEARCH_TRAILER_ERROR'
export const CLOSE_MODAL = 'CLOSE_MODAL'
