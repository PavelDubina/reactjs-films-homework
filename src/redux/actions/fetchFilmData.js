/*
 * Copyright Â© 2020 EPAM Systems, Inc. All Rights Reserved. All information contained herein is, and remains the
 * property of EPAM Systems, Inc. and/or its suppliers and is protected by international intellectual
 * property law. Dissemination of this information or reproduction of this material is strictly forbidden,
 * unless prior written permission is obtained from EPAM Systems, Inc
 */

import { SEARCH_DETAILS_SUCCESS, SEARCH_DETAILS_ERROR } from './actionTypes'
import { API_KEY, BASE_URL } from '../../constants/credentials'

export const fetchDataSuccess = (payload) => ({
  type: SEARCH_DETAILS_SUCCESS,
  payload,
})

export const fetchDataError = (error) => ({
  type: SEARCH_DETAILS_ERROR,
  payload: error,
})

export const fetchFilmData = (id) => async (dispatch) => {
  try {
    const responce = await fetch(`${BASE_URL}/3/movie/${id}?api_key=${API_KEY}`)
    const film = await responce.json()
    dispatch(fetchDataSuccess(film))
  } catch (error) {
    dispatch(fetchDataError(error))
  }
}
