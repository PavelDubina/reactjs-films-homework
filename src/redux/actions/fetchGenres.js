/*
 * Copyright Â© 2020 EPAM Systems, Inc. All Rights Reserved. All information contained herein is, and remains the
 * property of EPAM Systems, Inc. and/or its suppliers and is protected by international intellectual
 * property law. Dissemination of this information or reproduction of this material is strictly forbidden,
 * unless prior written permission is obtained from EPAM Systems, Inc
 */
import { SEARCH_GENRES_SUCCESS, SEARCH_GENRES_ERROR } from './actionTypes'
import { API_KEY } from '../../utils/constants'

const fetchDataSuccess = (payload) => ({
  type: SEARCH_GENRES_SUCCESS,
  payload,
})

const fetchDataError = (error) => ({
  type: SEARCH_GENRES_ERROR,
  payload: error,
})

const fetchGenresData = () => async (dispatch) => {
  try {
    const responce = await fetch(`https://api.themoviedb.org/3/genre/movie/list?api_key=${API_KEY}&language=en-US`)
    const genresList = await responce.json()
    dispatch(fetchDataSuccess(genresList.genres))
  } catch (error) {
    dispatch(fetchDataError(error))
  }
}

export default fetchGenresData
