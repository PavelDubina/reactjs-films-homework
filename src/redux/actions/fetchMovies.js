/*
 * Copyright Â© 2020 EPAM Systems, Inc. All Rights Reserved. All information contained herein is, and remains the
 * property of EPAM Systems, Inc. and/or its suppliers and is protected by international intellectual
 * property law. Dissemination of this information or reproduction of this material is strictly forbidden,
 * unless prior written permission is obtained from EPAM Systems, Inc
 */
import { SEARCH_MOVIES_PENDING, SEARCH_MOVIES_SUCCESS, SEARCH_MOVIES_ERROR } from './actionTypes'
import { API_KEY, BASE_URL } from '../../constants/credentials'

const fetchData = () => ({
  type: SEARCH_MOVIES_PENDING,
})

const fetchDataSuccess = (payload) => ({
  type: SEARCH_MOVIES_SUCCESS,
  payload,
})

const fetchDataError = (error) => ({
  type: SEARCH_MOVIES_ERROR,
  payload: error,
})

const fetchMoviesData = ({ search = false, category = '', genre = '', query = '', page = 1 }) => {
  const genreUrl = genre ? `&with_genres=${genre}` : ''
  const searchUrl = search ? `/search` : ''
  const queryUrl = query ? `&query=${query}` : ''
  const categoryUrl = category ? `/${category}` : ''
  return async (dispatch) => {
    try {
      dispatch(fetchData())
      const responce = await fetch(
        `${BASE_URL}/3${searchUrl}/movie${categoryUrl}?api_key=${API_KEY}&page=${page}${genreUrl}${queryUrl}`,
      )
      const movies = await responce.json()
      dispatch(fetchDataSuccess(movies.results))
    } catch (error) {
      dispatch(fetchDataError(error))
    }
  }
}

export default fetchMoviesData
