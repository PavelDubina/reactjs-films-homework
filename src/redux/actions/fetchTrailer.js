/*
 * Copyright Â© 2020 EPAM Systems, Inc. All Rights Reserved. All information contained herein is, and remains the
 * property of EPAM Systems, Inc. and/or its suppliers and is protected by international intellectual
 * property law. Dissemination of this information or reproduction of this material is strictly forbidden,
 * unless prior written permission is obtained from EPAM Systems, Inc
 */

import { SEARCH_TRAILER_SUCCESS, SEARCH_TRAILER_ERROR, CLOSE_MODAL } from './actionTypes'
import { API_KEY } from '../../utils/constants'

const fetchDataSuccess = (payload) => ({
  type: SEARCH_TRAILER_SUCCESS,
  payload,
})

const fetchDataError = (error) => ({
  type: SEARCH_TRAILER_ERROR,
  payload: error,
})

export const closeModal = () => ({
  type: CLOSE_MODAL,
})

export const fetchTrailerData = (id) => async (dispatch) => {
  try {
    const responce = await fetch(
      // eslint-disable-next-line max-len
      `https://api.themoviedb.org/3/movie/${id}/videos?api_key=${API_KEY}&language=en-US`,
    )
    const film = await responce.json()
    dispatch(fetchDataSuccess(film))
  } catch (error) {
    dispatch(fetchDataError(error))
  }
}
