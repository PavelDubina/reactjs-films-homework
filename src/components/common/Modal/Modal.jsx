/*
 * Copyright Â© 2020 EPAM Systems, Inc. All Rights Reserved. All information contained herein is, and remains the
 * property of EPAM Systems, Inc. and/or its suppliers and is protected by international intellectual
 * property law. Dissemination of this information or reproduction of this material is strictly forbidden,
 * unless prior written permission is obtained from EPAM Systems, Inc
 */

import React from 'react'
import { useSelector, useDispatch } from 'react-redux'
import ReactDOM from 'react-dom'
import YouTube from 'react-youtube'
import { closeModal } from '../../../redux/actions/fetchTrailer'
import styles from './Modal.scss'

const Modal = () => {
  const dispatch = useDispatch()
  const { open, trailer } = useSelector((state) => state.trailer)
  const opts = {
    height: '390',
    width: '640',
    playerVars: {
      autoplay: 1,
    },
  }
  return (
    open &&
    ReactDOM.createPortal(
      <div onClick={() => dispatch(closeModal())} className={styles.modal}>
        <YouTube videoId={trailer.results[0].key} opts={opts} />
      </div>,
      document.querySelector('#modal-root'),
    )
  )
}

export default Modal
