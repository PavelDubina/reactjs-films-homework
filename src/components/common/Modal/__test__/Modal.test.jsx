/*
 * Copyright Â© 2020 EPAM Systems, Inc. All Rights Reserved. All information contained herein is, and remains the
 * property of EPAM Systems, Inc. and/or its suppliers and is protected by international intellectual
 * property law. Dissemination of this information or reproduction of this material is strictly forbidden,
 * unless prior written permission is obtained from EPAM Systems, Inc
 */

import React from 'react'
import { create, act } from 'react-test-renderer'
import { BrowserRouter as Router } from 'react-router-dom'
import ReactDOM from 'react-dom'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import { resetTrailer } from '../../../../redux/actions/fetchTrailer'
import { Modal } from '../Modal'

const mockStore = configureStore([])

describe('Modal', () => {
  let store
  beforeEach(() => {
    store = mockStore({
      trailer: { key: 'key', error: null },
    })
  })
  beforeAll(() => {
    ReactDOM.createPortal = jest.fn((element) => element)
  })

  it('Should render and match snapshot with error', () => {
    store = mockStore({
      trailer: { key: '', error: 'error' },
    })
    const component = create(
      <Provider store={store}>
        <Router>
          <Modal />
        </Router>
      </Provider>,
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
  it('Should render and match snapshot with key', () => {
    const component = create(
      <Provider store={store}>
        <Router>
          <Modal />
        </Router>
      </Provider>,
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
  it('should dispatch an action on div click', () => {
    store.dispatch = jest.fn()
    const component = create(
      <Provider store={store}>
        <Router>
          <Modal />
        </Router>
      </Provider>,
    )
    act(() => {
      component.root.findAllByType('div')[0].props.onClick()
    })
    expect(store.dispatch).toHaveBeenCalledTimes(1)
    expect(store.dispatch).toHaveBeenCalledWith(resetTrailer())
  })
})
