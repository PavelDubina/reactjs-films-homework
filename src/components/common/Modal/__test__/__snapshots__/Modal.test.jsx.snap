// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Should render and match snapshot with error 1`] = `
<div
  className="modal"
  onClick={[Function]}
>
  <div
    className="not_found"
  >
    Trailer not found
  </div>
</div>
`;

exports[`Modal Should render and match snapshot with key 1`] = `
<div
  className="modal"
  onClick={[Function]}
>
  <iframe
    allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
    allowFullScreen={true}
    className="frame"
    frameBorder="0"
    src="https://www.youtube.com/embed/key?autoplay=1"
    title="youtube"
  />
</div>
`;
